ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestGeoResourceRecordSetApi.java

```diff
@@ -0,0 +1,413 @@
+      Iterator<ResourceRecordSet<?>> iterator = iteratorForDNameAndDirectionalType(name, dirType);
+      // retain original type (this will filter out CNAMEs)
+      return filter(iterator, filter);
+    } else {
+      return iteratorForDNameAndDirectionalType(name, dirType(type));
+    }
+  }
+
+  @Override
+  public ResourceRecordSet<?> getByNameTypeAndQualifier(String name, String type,
+                                                        String qualifier) {
+    checkNotNull(name, "description");
+    checkNotNull(type, "type");
+    checkNotNull(qualifier, "qualifier");
+    if (!supportedTypes.contains(type)) {
+      return null;
+    }
+    Iterator<DirectionalRecord> records = recordsByNameTypeAndQualifier(name, type, qualifier);
+    return nextOrNull(iteratorFactory.create(records, zoneName));
+  }
+
+  private Iterator<DirectionalRecord> recordsByNameTypeAndQualifier(String name, String type,
+                                                                                 String qualifier) {
+    if ("CNAME".equals(type)) {
+      return filter(
+          concat(recordsForNameTypeAndQualifier(name, "A", qualifier),
+                 recordsForNameTypeAndQualifier(name, "AAAA", qualifier)), isCNAME);
+    } else {
+      return recordsForNameTypeAndQualifier(name, type, qualifier);
+    }
+  }
+
+  private Iterator<DirectionalRecord> recordsForNameTypeAndQualifier(String name, String type,
+                                                                                  String qualifier) {
+    try {
+      return api.getDirectionalDNSRecordsForHost(zoneName, name, dirType(type))
+              .getDirectionalRecordsByGroup(qualifier).iterator();
+    } catch (UltraDNSRestException e) {
+      switch (e.code()) {
+        case UltraDNSRestException.GROUP_NOT_FOUND:
+        case UltraDNSRestException.DIRECTIONALPOOL_NOT_FOUND:
+          return Collections.<DirectionalRecord>emptyList().iterator();
+      }
+      throw e;
+    }
+  }
+
+  @Override
+  public void put(ResourceRecordSet<?> rrset) {
+    checkNotNull(rrset, "rrset was null");
+    checkArgument(rrset.qualifier() != null, "no qualifier on: %s", rrset);
+    checkArgument(IS_GEO.apply(rrset), "%s failed on: %s", IS_GEO, rrset);
+    checkArgument(supportedTypes.contains(rrset.type()), "%s not a supported type for geo: %s",
+            rrset.type(), supportedTypes);
+
+    int ttlToApply = rrset.ttl() != null ? rrset.ttl() : DEFAULT_TTL;
+    String group = rrset.qualifier();
+    Map<String, Collection<String>> regions = rrset.geo().regions();
+
+    DirectionalGroup directionalGroup = new DirectionalGroup();
+    directionalGroup.setName(group);
+    directionalGroup.setRegionToTerritories(regions);
+
+    List<Map<String, Object>> recordsLeftToCreate = new ArrayList<Map<String, Object>>(rrset.records());
+    Iterator<DirectionalRecord> iterator = recordsByNameTypeAndQualifier(rrset.name(), rrset.type(), group);
+
+    while (iterator.hasNext()) {
```

Comment:
I think I have already clarified on the mail that you don't need this logic.
you don’t need the logic to determine which record to add, update or delete for an existing owner.
Just call the PUT of REST & that will automatically take care of this.

So, the high-level logic for the implementation should look something like below:
- Call GET end-point for the input owner, zone & type combination
- If any rrset is returned:
  - Call PUT end-point to update (you already have all the information in the input parameter). Just use the input parameter to map it to the REST input for PUT
- If no rrset is returned:
  - Call POST end-point to create (you already have all the information in the input parameter). Just use the input parameter to map it to the REST input for POST

Please also manually run an integration test on the existing SOAP module & the suggested REST module code to verify that the end results are same.